### PR TITLE
TSDK-820 Refactor for descentralization

### PR DESCRIPTION
**Pull request by mundacho**
## Purpose

The goal of this task is to refactor the code to proceed with the decentralization of the bridge. 

## Approach

To accomplish this we did the following:

- split the project in two components: `consensus` and `public-api`.
- The `consensus` module will handle the BFT consensus.
- The `public-api` module will provide the WS to the clients of the services while also functioning as the client to the consensus layer.
- We implemented communications using gRPC.
- We implemented configuration files to configure the new components.
- We modified the integration tests to support the new architecture.
- Included @SeanCheatham PR to fix the release process.

## Testing

- The tests were fixed to support the new architecture.

## Tickets
* TSDK-812 TSDK-820

### CODE REVIEW

File: integration/src/test/scala/co/topl/bridge/FailedPeginNoDepositWithReorgModule.scala [hunk 1]
```diff
@@ -22,226 +10,54 @@ trait FailedPeginNoDepositWithReorgModule {
...
-        _ <- process
-          .ProcessBuilder(DOCKER_CMD, generateToAddress(1, 2, newAddress): _*)
-          .spawn[IO]
-          .use(_.exitValue)
-        _ <- process
-          .ProcessBuilder(DOCKER_CMD, generateToAddress(2, 8, newAddress): _*)
-          .spawn[IO]
-          .use(_.exitValue)
+        _ <- warn"We are in the waiting for escrow confirmation state"
+        // _ <- generateToAddress(1, 2, newAddress)
```
Comment (DiademShoukralla) [inline integration/src/test/scala/co/topl/bridge/FailedPeginNoDepositWithReorgModule.scala:44]:
```suggestion
```

File: integration/src/test/scala/co/topl/bridge/FailedPeginNoMint.scala [hunk 2]
```diff
@@ -1,140 +1,45 @@
...
-          .use(getText)
-        _ <- IO.println("newAddress: " + newAddress)
-        unspent <- process
-          .ProcessBuilder(DOCKER_CMD, extractGetTxId: _*)
-          .spawn[IO]
-          .use(getText)
-        // _ <- IO.println("unspent: " + unspent)
-        txId <- IO.fromEither(
-          parse(unspent).map(x => (x \\ "txid").head.asString.get)
+        // ref <- Ref.of[IO, Int](0)
```
Comment (DiademShoukralla) [inline integration/src/test/scala/co/topl/bridge/FailedPeginNoMint.scala:16]:
```suggestion
```

File: public-api/src/main/scala/co/topl/bridge/publicapi/ConsensusClientGrpc.scala [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        ConsensusClientMessageId,
+        Ref[F, Either[
+          BridgeError,
+          BridgeResponse
+        ]]
+      ],
+      channel: ManagedChannel
+      // address: String,
+      // port: Int,
+      // secureConnection: Boolean
```
Comment (DiademShoukralla) [inline public-api/src/main/scala/co/topl/bridge/publicapi/ConsensusClientGrpc.scala:59]:
```suggestion
```

File: public-api/src/main/scala/co/topl/bridge/publicapi/ConsensusClientGrpc.scala [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+  ) = {
+
+    // val channel = ManagedChannelBuilder
+    //   .forAddress(address, port)
+    for {
+      // channel <-
+      //   (if (secureConnection) channel.useTransportSecurity()
+      //    else channel.usePlaintext()).resource[F]
```
Comment (DiademShoukralla) [inline public-api/src/main/scala/co/topl/bridge/publicapi/ConsensusClientGrpc.scala:67]:
```suggestion
    for {
```

File: public-api/src/main/scala/co/topl/bridge/publicapi/controllers/StartSessionController.scala
Comment (DiademShoukralla) [inline public-api/src/main/scala/co/topl/bridge/publicapi/controllers/StartSessionController.scala:1]: Was this file meant to be committed?

File: public-api/src/main/scala/co/topl/bridge/publicapi/modules/ReplyServicesModule.scala [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            _ <- trace"Signature is valid"
+            _ <-
+              if (isValidSignature) {
+                val response: Either[BridgeError, BridgeResponse] =
+                  request.result match {
+                    case StateMachineReply.Result.Empty =>
+                      Left(
+                        shared.UnknownError("This should not happen: Empty response")
+                      )
+                    case MintingStatus(value) => // FIXME: Implement this
```
Comment (DiademShoukralla) [inline public-api/src/main/scala/co/topl/bridge/publicapi/modules/ReplyServicesModule.scala:71]: For my own understanding, what is left to be implemented here?

File: topl-btc-cli/src/main/scala/co/topl/tbcli/Main.scala [hunk 6]
```diff
@@ -3,73 +3,65 @@ package co.topl.tbcli
...
+    // OParser.parse(parser, args, ToplBTCCLIParamConfig()) match {
+    //   case Some(config) =>
+    //     config.command match {
+    //       case Some(command) =>
+    //         processCommand(config.btcNetwork, command)
+    //       case None =>
+    //         IO.consoleForIO.errorln("No command specified") *>
+    //           IO(ExitCode.Error)
+    //     }
+    //   case None =>
```
Comment (DiademShoukralla) [inline topl-btc-cli/src/main/scala/co/topl/tbcli/Main.scala:19]:
```suggestion
```

+ the other commented out sections in this file

File: topl-btc-cli/src/main/scala/co/topl/tbcli/TBCLIParamsDescriptor.scala
Comment (DiademShoukralla) [inline topl-btc-cli/src/main/scala/co/topl/tbcli/TBCLIParamsDescriptor.scala:1]: Is this file still needed?

File: topl-btc-cli/src/main/scala/co/topl/tbcli/TBCliParamConfig.scala [hunk 7]
```diff
@@ -1,6 +1,6 @@
 package co.topl.tbcli
-import co.topl.shared.BitcoinNetworkIdentifiers
-import co.topl.shared.RegTest
+// import co.topl.bridge.BitcoinNetworkIdentifiers
+// import co.topl.bridge.RegTest
```
Comment (DiademShoukralla) [inline topl-btc-cli/src/main/scala/co/topl/tbcli/TBCliParamConfig.scala:3]:
```suggestion
```

**Review by DiademShoukralla — APPROVED** [review]
LGTM! Just pointed out some areas for code cleanup, which you can ignore if they were left there for a reason.

File: consensus/src/main/scala/co/topl/bridge/consensus/statemachine/pegin/PeginStateMachine.scala [hunk 8]
```diff
@@ -111,6 +114,8 @@ object PeginStateMachine {
           fs2.Stream(
             for {
               x <- currentToplNetworkHeight.get
+              _ <- trace"current topl height is $x"
+              _ <- trace"Updating topl height to $height"
               _ <-
                 if (height > x) // TODO: handle reorgs
```
[the remark this answers is not in the document]
Comment (mundacho) [reply consensus/src/main/scala/co/topl/bridge/consensus/statemachine/pegin/PeginStateMachine.scala:120]: No, fixed

File: integration/prepareAndLauncIntegrationTest.sh [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+brambl-cli simple-minting create --from-fellowship self --from-template default  -h localhost --port 9084 -n private --keyfile $TOPL_WALLET_JSON -w $TOPL_WALLET_PASSWORD -o seriesMintingTx.pbuf -i seriesPolicy.yaml  --mint-amount 1 --fee 10 --walletdb $TOPL_WALLET_DB --mint-token series
+brambl-cli tx prove -i seriesMintingTx.pbuf --walletdb $TOPL_WALLET_DB --keyfile $TOPL_WALLET_JSON -w $TOPL_WALLET_PASSWORD -o seriesMintingTxProved.pbuf
+export ASSET_UTXO=$(brambl-cli tx broadcast -i seriesMintingTxProved.pbuf -h 127.0.0.1 --port 9084)
+brambl-cli bifrost-query mint-block --nb-blocks 1 -h 127.0.0.1  --port 9084 -s false
+echo "ASSET_UTXO: $ASSET_UTXO"
+until brambl-cli genus-query utxo-by-address --host localhost --port 9084 --secure false --walletdb $TOPL_WALLET_DB; do sleep 5; done
+
+# Run the tests
+# echo "Running integration tests"
+# sbt "integration/test"
```
[the remark this answers is not in the document]
Comment (mundacho) [reply integration/prepareAndLauncIntegrationTest.sh:102]: commented only, thanks

File: public-api/src/main/scala/co/topl/bridge/publicapi/modules/ReplyServicesModule.scala (hunk 5 above)
Comment (mundacho) [reply public-api/src/main/scala/co/topl/bridge/publicapi/modules/ReplyServicesModule.scala:71]: I had forgotten to remove the comment.

File: topl-btc-cli/src/main/scala/co/topl/tbcli/Main.scala (hunk 6 above)
Comment (mundacho) [reply topl-btc-cli/src/main/scala/co/topl/tbcli/Main.scala:19]: This whole module will disappear, but for the moment, we are keeping it until the wallet is done. I keep the code commented in case I need it again.

File: topl-btc-cli/src/main/scala/co/topl/tbcli/TBCLIParamsDescriptor.scala
Comment (mundacho) [reply topl-btc-cli/src/main/scala/co/topl/tbcli/TBCLIParamsDescriptor.scala:1]: same as before, will be removed with the module.
